Allow HPOS screens to work for custom order types

jorgeatorres opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:
<!-- Describe the changes made to this Pull Request and the reason for such changes. -->
Adds a few new APIs mainly to `ListTable` and `PageController` to support custom order types.
It also enhances `wc_get_page_screen_id()` so that it returns screen IDs for custom order types too.

The PR also updates some strings in the UI that were hardcoded with "Orders" or "Edit Order" for the corresponding label taken from the WordPress post type.

Most of this work was done in order to support Subscriptions but it benefits custom order types in general.

Closes #35601.

<!-- The next section is mandatory. If your PR doesn't require testing, please indicate that you are purposefully omitting instructions. -->

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Otherwise, please include detailed instructions on how these changes can be tested (including pre-conditions, configuration, steps to take and expected results). It may help to write your instructions using pseudocode -- as if you're telling a computer how to execute the test. -->

#### Test general support for custom types
1. Start with HPOS disabled. That is, "High performance order storage" disabled in WC > Settings > Features.
1. Add and activate the following snippet either as a standalone plugin (or add to `functions.php` or similar). The job of this snippet is to create a custom order type so that there's actually something to test.
   <details>
     <summary>(PHP snippet)</summary>

    ```php
    add_action(
	    'plugins_loaded',
	    function() {
    
		    class WC_Shop_Ticket extends \WC_Order {
    
			    public function get_type() {
				    return 'shop_ticket';
			    }
    
		    }
    
		    add_action(
			    'init',
			    function() {
				    wc_register_order_type(
					    'shop_ticket',
					    array(
						    'labels'                           => array(
							    'name'               => __( 'Tickets', 'cust-ord-typ' ),
							    'singular_name'      => __( 'Ticket', 'cust-ord-typ' ),
							    'add_new'            => _x( 'Add Ticket', 'custom post type setting', 'cust-ord-typ' ),
							    'add_new_item'       => _x( 'Add New Ticket', 'custom post type setting', 'cust-ord-typ' ),
							    'edit'               => _x( 'Edit', 'custom post type setting', 'cust-ord-typ' ),
							    'edit_item'          => _x( 'Edit Ticket', 'custom post type setting', 'cust-ord-typ' ),
							    'new_item'           => _x( 'New Ticket', 'custom post type setting', 'cust-ord-typ' ),
							    'view'               => _x( 'View Ticket', 'custom post type setting', 'cust-ord-typ' ),
							    'view_item'          => _x( 'View Ticket', 'custom post type setting', 'cust-ord-typ' ),
							    'search_items'       => __( 'Search Tickets', 'cust-ord-typ' ),
							    'not_found'          => __( 'Tickets Not Found', 'cust-ord-typ' ),
							    'not_found_in_trash' => _x( 'No tickets found in trash', 'custom post type setting', 'cust-ord-typ' ),
							    'parent'             => _x( 'Parent Tickets', 'custom post type setting', 'cust-ord-typ' ),
							    'menu_name'          => __( 'Tickets', 'cust-ord-typ' ),
						    ),
						    'description'                      => __( 'This is where tickets are stored.', 'cust-ord-typ' ),
						    'public'                           => false,
						    'show_ui'                          => true,
						    'capability_type'                  => 'shop_order',
						    'map_meta_cap'                     => true,
						    'publicly_queryable'               => false,
						    'exclude_from_search'              => true,
						    'show_in_menu'                     => current_user_can( 'manage_woocommerce' ) ? 'woocommerce' : true,
						    'hierarchical'                     => false,
						    'show_in_nav_menus'                => false,
						    'rewrite'                          => false,
						    'query_var'                        => false,
						    'supports'                         => array( 'title', 'comments', 'custom-fields' ),
						    'has_archive'                      => false,
    
						    // wc_register_order_type() params
						    'exclude_from_orders_screen'       => true,
						    'add_order_meta_boxes'             => true,
						    'exclude_from_order_count'         => true,
						    'exclude_from_order_views'         => true,
						    'exclude_from_order_webhooks'      => true,
						    'exclude_from_order_reports'       => true,
						    'exclude_from_order_sales_reports' => true,
						    'class_name'                       => 'WC_Shop_Ticket',
					    )
				    );
			    }
		    );
    
	    }
    );
    ```
   </details>
1. With the above snippet enabled, you should see a new "Tickets" menu item under the WooCommerce menu.
   Notice this is a regular WP post type (i.e. the URL is `edit.php?post_type=shop_ticket`).
1. Go to WC > Tickets and create a few tickets for testing purposes.
1. Now go to WC > Settings > Features and enable HPOS.
1. Make sure to enable sync so that the created tickets are migrated over to HPOS.
1. Confirm the "Tickets" menu item still appears under the WooCommerce menu, but now its URL no longer refers to a custom post type URL. The new one should be `admin.php?page=wc-orders--shop_ticket`.
1. Confirm the tickets you created in step 4 are present.
1. Confirm you can add or edit tickets.
1. Check that the labels around the UI use the correct text, such as "Add Ticket" or "Edit Ticket".

<!-- End testing instructions -->

### Other information:

-   [X] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [X] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

jorgeatorres committed:
Allow `wc_get_order_types()` to return list of order types that have UI

jorgeatorres committed:
Make `PageController` register menu items for all order types required

jorgeatorres committed:
Make `PageController` URL generation aware of order types

jorgeatorres committed:
Make `Edit` order type aware

jorgeatorres committed:
Add support to `wc_get_page_screen_id()` for order types

jorgeatorres committed:
Make `PageController` order type aware

jorgeatorres committed:
Include custom order type screens as WC screens

jorgeatorres committed:
Make `get_base_page_url()` in `PageController` public

jorgeatorres committed:
Make sure Trash links use the correct URL

jorgeatorres committed:
Gather request vars in `$request` in list table

jorgeatorres committed:
Add some order type specific filters to ListTable

jorgeatorres committed:
Fix wrong reference to orders per page in ListTable

jorgeatorres committed:
Make ListTable columns work in all screens

jorgeatorres committed:
Make PHPCS happy

jorgeatorres committed:
Minor PHPCS fix

jorgeatorres committed:
Add changelog

haszari commented:
Looks like a bunch of e2e tests are failing, not clear if these are due to this PR or unrelated. Will investigate 🕵️

brucealdridge committed:
Set page hidden form field based on the order_type (#35751)

jorgeatorres committed:
Add ‘woocommerce_order_list_table_extra_tablenav’ hook

jorgeatorres committed:
Fix PHPCS warnings

jorgeatorres committed:
Make `wc_get_page_screen_id()` datastore aware

jorgeatorres committed:
Make `WC_Admin_Assets::is_order_meta_box_screen()` aware of HPOS screen IDs

jorgeatorres committed:
Remove unnecessary `$context` arg from `wc_get_page_screen_id()`

jorgeatorres committed:
Add support for custom order types to `PostsRedirectionController`

jorgeatorres committed:
Make PHPCS happy

jorgeatorres committed:
Only connect COT admin page when COT is enabled

vedanshujain commented on plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php:
```diff
@@ -350,27 +407,35 @@ private function set_customer_args() {
-			$query_statuses = array_intersect(
-				array_keys( wc_get_order_statuses() ),
-				get_post_stati( array( 'show_in_admin_all_list' => true ), 'names' )
+		if ( empty( $status ) || in_array( 'all', $status, true ) ) {
```
note that this is never empty, because of `array_map` this will be an array of at least one element whereas earlier it would have been a string. therefore the logic inside is never executed even when the request['status'] is emtpy

jorgeatorres replied:
You're right. It's been fixed in d7965ac3d109c8797f7e2e62eb16261bacf43612.

vedanshujain reviewed (changes requested):
Requesting a bug fix, otherwise looks good.

jorgeatorres committed:
Properly filter out empty status in ListTable

jorgeatorres commented:
Hi @vedanshujain!

Thanks for the review 🙇‍♂️. I've pushed a fix for the issue you raised. Let me know what you think!

jorgeatorres committed:
Make `ListTable` order type aware

jorgeatorres committed:
Fix warning due to uninitialized var

jorgeatorres committed:
Do not rely on `wc_orders_count()` for counting orders in `ListTable`

`wc_orders_count()` assumes that order types correspond to a datastore’s object type, which isn’t necessarily the case, producing incorrect results.
Until this is addressed, it’d be best not to rely on it for types other than orders.

jorgeatorres commented:
@vedanshujain:

After #35370 was merged I noticed some conflicts in this PR. The conflicts weren't serious, but I also noticed that relying on `wc_orders_count()` (even the modified version introduced in #35370) broke things for this PR, producing some fatal errors due to an "Invalid datastore".

That was due to how `wc_orders_count()` [loads the datastore for a given order type](https://github.com/woocommerce/woocommerce/blob/51c33fa351b588ea9d0b4c3bf53025faec3cde02/plugins/woocommerce/includes/wc-order-functions.php#L366
), assuming it has the same name as the order type itself, which isn't the case most of the time. Core itself uses `order` for the datastore name but the type is `shop_order`. Subscriptions, for example, registers the datastore with object type `'subscription'`, but the order type is actually `shop_subscription`. Something similar happens in Bookings (`booking` vs `wc_booking`).

This was even the case before #35370, so we probably never noticed this because most order types are registered with the `exclude_from_order_count` flag set to `true`.

Given the urgency of merging this (and the possibility of later revisiting the issues I mentioned) I decided to go with a more manual approach for counting orders in the list table: 0081806836e5ec90b33b36fd7b11e89cafe29dda introduces a direct query, which ironically, was exactly what I was trying to avoid in #35370.

Please let me know what you think, and if you think we should create an issue to investigate or try to address the discrepancy between order types/datastore object types in general.

vedanshujain commented on plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php:
```diff
@@ -105,15 +125,28 @@ public function render_column( $column_id, $order ) {
-		 * @since 7.0.0
-		 */
-		do_action( "manage_{$this->screen->id}_custom_column", $column_name, $order );
+		if ( has_action( 'woocommerce_' . $this->order_type . '_list_table_custom_column' ) ) {
```
won't they both need to be fired? for example, since it only checked for the filter name, an action for one column (say column1) can disable action for another column (say column2), if column1 is registering on `woocommerce_{$this->order_type}_list_table_column` and column2 is registering on `manage_{$this->screen_id}_custom_column`.

jorgeatorres replied:
Yes. I would've preferred we don't have this "generic" version of the hook (instead of the order-type specific one), but given it exists, I agree that it'd be best to fire both just in case. I've done that in cb6a787c59c982c4b95624d314943f703730b72d.

vedanshujain commented on plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php:
```diff
@@ -124,7 +157,7 @@ private function items_per_page(): void {
 			array(
 				'default' => 20,
-				'option'  => 'edit_orders_per_page',
+				'option'  => 'edit_' . $this->order_type . '_per_page',
```
good catch!

vedanshujain commented on plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php:
```diff
@@ -421,14 +486,34 @@ public function get_views() {
+
+		$status_placeholders = implode( ',', array_fill( 0, count( $status ), '%s' ) );
+
+		$count = absint(
```
we can improve this query by using a group by here and doing it for all statuses at once instead of firing this query again and again. since this is a private method, it should be fine to change the signature, wdyt? as far as I can see, a group by query on status should be similarly fast to a normal count query for a single status (since MYSQL would need to scan entire `status` index anyway) making us overall N times faster where N is number of status.

jorgeatorres replied:
I didn't actually change the signature, but the first time the function is called it computes counts for all statuses and then caches the result. It's implemented in b98a1660b4de8fee24082937a986568455e1f72f. Let me know what you think 😃

vedanshujain reviewed (commented):
Added couple of comments, looks good otherwise. Great work!

jorgeatorres committed:
Improve performance for status counts in ListTable

jorgeatorres committed:
Fire both order-type specific and generic hook for default column in ListTable

jorgeatorres commented:
@vedanshujain: Thanks for the feedback. I've implemented your suggestions and have requested a re-review. Let me know what you think!

vedanshujain reviewed (approved):
LGTM!
